shpc/main/modules/__init__.py

```diff
@@ -409,43 +396,38 @@ def write_version_file(self, uri, tag, default_choice=False):
 
         # First or last installed
         elif self.settings.default_version == "last_installed":
-            self._default_version_last_installed(version_file, tag)
+            self._set_default_version(version_file, tag)
 
         elif self.settings.default_version == "first_installed":
-            self._default_version_first_installed(version_file, tag, default_choice)
+            first_installed = True if len(os.listdir(version_dir)) == 1 else False
```

Comment:
This is going to be wrong given an empty directory with just .version. Also given we do this more than once, should we make a common helper function?

Reply:
Absolutely, and that's another reason why I've introduced the other `_cleanup` above. Note that I didn't touch the logic `True if len(os.listdir(version_dir)) == 1 else False`, just moved it from another function to here, so the weakness you're describing was already there.

Reply:
(continued)
So: in theory if the directory exists and has exactly 1 entry, it should be an actual software version. But to be safe I can add a helper function if you prefer ?

Reply:
My suggestion is to fix the cleanup function to handle the .version file (and nothing else) and then visit this again.